packages/atomic/src/autoloader/index.ts

```diff
@@ -22,11 +22,18 @@ export function registerAutoloader(
    */
   const observeStencilElementHydration = (atomicElement: Element) => {
     const attributeObserver = new MutationObserver(async () => {
-      if (atomicElement.classList.contains('hydrated')) {
+      if (
+        atomicElement.classList.contains('hydrated') &&
+        'shadowRoot' in atomicElement &&
+        atomicElement.shadowRoot &&
+        !visitedNodes.has(atomicElement.shadowRoot)
+      ) {
         attributeObserver.disconnect();
-        if ('shadowRoot' in atomicElement && atomicElement.shadowRoot) {
-          await discover(atomicElement.shadowRoot);
-        }
```

Comment:
essentially:
1. Coalesce all conditions in a single if
2. When all conditions are met: discover whatever is the shadowDOM, observe it, and stop observing the attribute of the element.